sled-agent/src/rack_setup/service.rs

```diff
@@ -51,176 +93,450 @@ impl Service {
-        // (destructive) initialization has occurred.
+        // Check if a previous RSS plan has completed successfully.
         //
-        // We do this by storing the configuration at "rss_config_path"
-        // after successfully performing initialization.
-        let rss_config_path =
-            std::path::Path::new(omicron_common::OMICRON_CONFIG_PATH)
-                .join("config-rss.toml");
-        if rss_config_path.exists() {
+        // If it has, the system should be up-and-running.
+        let rss_completed_plan_path = rss_completed_plan_path();
+        if rss_completed_plan_path.exists() {
+            // TODO(https://github.com/oxidecomputer/omicron/issues/724): If the
+            // running configuration doesn't match Config, we could try to
+            // update things.
             info!(
                 self.log,
-                "RSS configuration already exists at {}",
-                rss_config_path.to_string_lossy()
+                "RSS configuration looks like it has already been applied",
             );
-            let old_config: Config = toml::from_str(
-                &tokio::fs::read_to_string(&rss_config_path).await?,
-            )?;
-            if &old_config == config {
-                info!(
-                    self.log,
-                    "RSS config already applied from: {}",
-                    rss_config_path.to_string_lossy()
-                );
-                return Ok(());
-            }
-
-            // TODO(https://github.com/oxidecomputer/omicron/issues/724):
-            // We could potentially handle this case by deleting all
-            // datasets (in preparation for applying the new
-            // configuration), but at the moment it's an error.
-            warn!(
-                self.log,
-                "Rack Setup Service Config was already applied, but has changed.
-                This means that you may have datasets set up on this sled, but they
-                may not match the ones requested by the supplied configuration.\n
-                To re-initialize this sled:
-                   - Disable all Oxide services
-                   - Delete all datasets within the attached zpool
-                   - Delete the configuration file ({})
-                   - Restart the sled agent",
-                rss_config_path.to_string_lossy()
-            );
-            return Err(SetupServiceError::Configuration);
+            return Ok(());
         } else {
-            info!(
-                self.log,
-                "No RSS configuration found at {}",
-                rss_config_path.to_string_lossy()
-            );
+            info!(self.log, "RSS configuration has not been fully applied yet",);
         }
 
+        // Wait for either:
+        // - All the peers to re-load an old plan (if one exists)
+        // - Enough peers to create a new plan (if one does not exist)
+        let maybe_plan = self.load_plan().await?;
```

Comment:
I only just now understood the naming of `PeerExpectation::Precise` and `PeerExpectation::Arbitrary`. If the intention is that we use the former iff we've already written the plan to disk, and the latter iff we've not, it might be helpful to rename them to reflect that. At least, a bit more documentation on those variants would be helpful.